src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy

```diff
@@ -1720,45 +1796,339 @@ class ActionDelegate {
+                        "type" : "text",
+                        "value": body.icon
+                ],
+                (PREFERENCE_ITEM_FIELD_NODE_PATH)    : [
+                        "type" : "text",
+                        "value": node.uriPath
                 ],
         ]
         setData(newItemTask, setDataMap)
         finishTask(newItemTask)
 
-        def task = orgGroup.tasks.find { it.transition == "append_menu_item" }.task
-        dataService.setData(task, ImportHelper.populateDataset([
-                (PREFERENCE_ITEM_FIELD_APPEND_MENU_ITEM): ["type": "text", "value": itemCase.stringId]
-        ]))
+        folder = workflowService.findOne(folder.stringId)
+        if (node.parentId != null) {
+            UriNode parentNode = uriService.findById(node.parentId)
+            body = new MenuItemBody(new I18nString(parentNode.name), "folder")
+
+            getOrCreateFolderRecursive(parentNode, body, folder)
+        }
+
+        return folder
+    }
+
+    void moveMenuItem(Case item, String destUri) {
+        if (isCyclicNodePath(item, destUri)) {
+            throw new IllegalArgumentException("Cyclic path not supported. Destination path: ${destUri}")
+        }
+
+        List<Case> casesToSave = new ArrayList<>()
+
+        String parentId = item.dataSet[PREFERENCE_ITEM_FIELD_PARENT_ID].value
+        if (parentId != null) {
+            Case oldParent = removeChildItemFromParent(parentId, item)
+            casesToSave.add(oldParent)
+        }
+
+        UriNode destNode = uriService.getOrCreate(destUri, UriContentType.CASE)
+        Case newParent = getOrCreateFolderItem(destNode.uriPath)
+        if (newParent != null) {
+            item.dataSet[PREFERENCE_ITEM_FIELD_PARENT_ID].value = newParent.stringId
+            newParent = appendChildCaseId(newParent, item.stringId)
+            casesToSave.add(newParent)
+        } else {
+            item.dataSet[PREFERENCE_ITEM_FIELD_PARENT_ID].value = null
+        }
+
+        item.uriNodeId = destNode.id
+        item = resolveAndHandleNewNodePath(item, destNode.uriPath)
+        casesToSave.add(item)
+
+        if (hasChildren(item)) {
+            List<Case> childrenToSave = updateNodeInChildrenFoldersRecursive(item)
+            casesToSave.addAll(childrenToSave)
+        }
+
+        for (aCase in casesToSave) {
+            if (aCase != null) {
+                workflowService.save(aCase)
+            }
+        }
+    }
+
+    Case duplicateMenuItem(Case originItem, I18nString newTitle, String newIdentifier = null) {
+        if (!newIdentifier) {
+            throw new IllegalArgumentException("View item identifier is null!")
+        }
+        String sanitizedIdentifier = sanitize(newIdentifier)
+        if (findMenuItem(sanitizedIdentifier)) {
+            throw new IllegalArgumentException("View item identifier $sanitizedIdentifier is not unique!")
+        }
+        if (newTitle == null || newTitle.defaultValue == "") {
```

Comment:
I would move this call right after `newIdentifier` check. If it can throw exception then calling `sanitize` and `findMenuItem` is unnecessary.